pyigm/cgm/models.py

```diff
@@ -0,0 +1,240 @@
+        # Cosmology
+        self.H0 = 70. *u.km/u.s/ u.Mpc
+        self.fb = 0.16       # Baryon fraction
+        self.rhoc = 9.2e-30 * u.g / u.cm**3
+        # Dark Matter
+        self.r200 = (((3*10**self.M_halo * const.M_sun.cgs) / (4*np.pi*200*self.rhoc))**(1/3)).to('kpc')
+        self.rho0 = 200*self.rhoc/3  * self.c**3 / self.fy(self.c)   # Central density
+        # Misc
+        self.mu = 1.33   # Reduced mass correction for Helium
+
+    def fy(self, y):
+        """ Enclosed mass function
+        Parameters
+        ----------
+        y : float or ndarray
+
+        Returns
+        -------
+        f_y : float or ndarray
+
+        """
+        f_y = np.log(1+y) - y/(1+y)
+        return f_y
+
+    def ne(self, xyz):
+        """ Calculate n_e from n_H with a correction for Helium
+        Assume 25% mass is Helium and both electrons are off
+
+        Parameters
+        ----------
+        xyz
+
+        Returns
+        -------
+        n_e : float or ndarray
+          electron density in cm**-3
+
+        """
+        ne = self.nH(xyz) * 1.1667
+        # Return
+        return ne
+
+    def nH(self, xyz):
+        """ Calculate the Hydrogen number density
+        Includes a correction for Helium
+
+        Parameters
+        ----------
+        xyz
+
+        Returns
+        -------
+        nH : float or ndarray
+          Density in cm**-3
+
+        """
+        nH = self.rho(xyz).value / self.mu / m_p
+        # Return
+        return nH
+
+    def rho(self, xyz):
+        """ Mass density in baryons; modified
+
+        Parameters
+        ----------
+        xyz : ndarray
+          Position assumed in kpc
+
+        Returns
+        -------
+        rho : Quantity
+          Density in g / cm**-3
+
+        """
+        radius = np.sqrt(rad3d2(xyz))
+        y = self.c * (radius/self.r200.to('kpc').value)
+        rho = (self.fb*self.f_hot*self.rho0) / y**(1-self.alpha) / (self.y0+y)**(2+self.alpha)
+        # Return
+        return rho
+
+    def Ne_Rperp(self, Rperp, step_size=0.1*u.kpc, rmax=1., epsrel=1e-4, epsabs=1e-6,
+                 *arg, **kwargs):
+        """ Calculate N_H at an input impact parameter Rperp
+
+        Parameters
+        ----------
+        Rperp : Quantity
+          Impact parameter, typically in kpc
+        step_size : Quantity
+          Step size used for numerical integration
+        rmax : float
+          Maximum radius for integration in units of r200
+
+        Returns
+        -------
+        Ne : Quantity
+          Column density of total electrons
+        """
+        dz = step_size.to('kpc').value
+
+        # Cut at rmax*rvir
+        if Rperp > rmax*self.r200:
+            return 0. / u.cm**2
+        # Generate a sightline to rvir
+        zmax = np.sqrt(self.r200 ** 2 - Rperp ** 2).to('kpc')
+        zval = np.arange(-zmax.value, zmax.value+dz, dz)  # kpc
+        # Set xyz
+        xyz = np.zeros((3,zval.size))
+        xyz[0, :] = Rperp.to('kpc').value
```

Comment:
@profxj (this is purely a suggestion): Here xyz[1, :]=0 since it's only a 1D profile along r, for future work, maybe it would be helpful to define a position angle _theta_, which could be used to define the x/y axis value of the sightline array on a polar coordinate system. In such case, xyz[0, :]=Rperp*np.cos(theta)+x0, xyz[1, :]=Rperp*np.sin(theta)+y0, where (x0, y0) is the origin of the coordinate system (see my next comment).